File: samples/nodejs/transactions.js
//
// transactions.js — MVCC isolation seen from the client (JavaScript twin of
// ../cpp/transactions_demo.cpp; see ../../transactions-and-concurrency.md).
//
// Two connections play the same scenario: SNAPSHOT stability, READ COMMITTED
// freshness, then a write conflict surfaced as SQLSTATE 40001.
//
// Run:  node transactions.js
//
'use strict';

const { attachOrCreate, attach, ISOLATION, s } = require('./common');

const DB = { database: process.argv[2] || '/tmp/fbhandson/tx.fdb' };

(async () => {
    const a = await attachOrCreate(DB);   // connection A
    const b = await attach(DB);           // connection B

    await a.query('recreate table balance (id integer primary key, amount integer)');
    await a.query('insert into balance values (1, 100)');

    // --- 1. SNAPSHOT stability ---------------------------------------------
    const snapA = await a.transaction(ISOLATION.SNAPSHOT);
    let r = await snapA.query('select amount from balance where id = 1');
    console.log('A (SNAPSHOT)       sees amount =', r[0].AMOUNT);

    await b.query('update balance set amount = 999 where id = 1');
    console.log('B                  committed amount = 999');

    r = await snapA.query('select amount from balance where id = 1');
    console.log('A (same SNAPSHOT)  sees amount =', r[0].AMOUNT,
        '  <- still the start-of-tx version');

    // --- 2. READ COMMITTED sees the new version ----------------------------
    const rcA = await a.transaction(ISOLATION.READ_COMMITTED);
    r = await rcA.query('select amount from balance where id = 1');
    console.log('A (READ COMMITTED) sees amount =', r[0].AMOUNT,
        '  <- the committed version');
    await rcA.commit();

    // --- 3. Write conflict -------------------------------------------------
    // B updates the row inside an open SNAPSHOT transaction; A, also under
    // SNAPSHOT, updates the same row and commits first.  When B's turn comes
    // the engine reports the update conflict (node-firebird's default TPB is
    // WAIT, so the conflict surfaces when the blocker commits).
    const holdB = await b.transaction(ISOLATION.SNAPSHOT);
    await holdB.query('update balance set amount = amount + 1 where id = 1');

    const loserA = await a.transaction(ISOLATION.SNAPSHOT);
    const race = loserA.query('update balance set amount = amount + 10 where id = 1')
        .then(() => 'unexpected: conflicting update succeeded')
        .catch(e => 'A conflicting update failed as designed:\n    ' + e.message);
    await new Promise(res => setTimeout(res, 300));   // let A block on the row
    await holdB.commit();                             // blocker commits -> conflict fires
    console.log(await race);

    await loserA.rollback();
    await snapA.commit();
    await a.detach();
    await b.detach();
    console.log('done.');
})().catch(e => { console.error(e); process.exit(1); });

File: samples/rust/src/bin/transactions.rs
//! transactions.rs — the transactions-and-concurrency scenario in Rust.
//!
//! The rsfbclient twin of ../cpp/transactions_demo.cpp: two attachments,
//! SNAPSHOT vs READ COMMITTED visibility of a concurrent commit, then a
//! NO WAIT update conflict.  The TPB decisions the OO-API sample spells
//! out byte by byte are a typed TransactionConfiguration here —
//! TrIsolationLevel::Concurrency / ReadCommited(RecordVersion),
//! TrLockResolution::NoWait — applied through explicit SimpleTransaction
//! objects.  (That explicitness matters: rsfbclient connections also carry
//! a hidden DEFAULT transaction whose commit()/rollback() are RETAINING —
//! it keeps its first configuration and, under SNAPSHOT, its first
//! snapshot, for the life of the connection.)
//! See ../../transactions-and-concurrency.md.
//!
//! Run (see ../README.md):  cargo run --bin transactions

use fb_handson_rust::connect;
use rsfbclient::{prelude::*, FbError, SimpleTransaction};

fn main() -> Result<(), FbError> {
    let mut a = connect("transactions")?;
    let mut b = connect("transactions")?;

    let snapshot = TransactionConfiguration {
        isolation: TrIsolationLevel::Concurrency,
        lock_resolution: TrLockResolution::NoWait,
        ..TransactionConfiguration::default()
    };
    let read_committed = TransactionConfiguration {
        isolation: TrIsolationLevel::ReadCommited(TrRecordVersion::RecordVersion),
        lock_resolution: TrLockResolution::NoWait,
        ..TransactionConfiguration::default()
    };

    // Scratch table (idempotent).
    {
        let mut setup = SimpleTransaction::new(&mut a, read_committed)?;
        let _ = setup.execute("DROP TABLE ACCOUNTS", ());
        setup.execute(
            "CREATE TABLE ACCOUNTS (ID INT NOT NULL PRIMARY KEY, BALANCE INT)",
            (),
        )?;
        setup.commit()?;
    }
    {
        let mut setup = SimpleTransaction::new(&mut a, read_committed)?;
        setup.execute("INSERT INTO ACCOUNTS VALUES (1, 100)", ())?;
        setup.commit()?;
    }

    // -- 1. Visibility: a SNAPSHOT transaction cannot see a commit that
    //       happens after it starts; a READ COMMITTED one can.
    let mut tr_a = SimpleTransaction::new(&mut a, snapshot)?;
    let (before,): (i64,) = tr_a
        .query_first("SELECT BALANCE FROM ACCOUNTS WHERE ID = 1", ())?
        .unwrap();
    println!("snapshot sees balance {} at start", before);

    {
        let mut tr_b = SimpleTransaction::new(&mut b, read_committed)?;
        tr_b.execute("UPDATE ACCOUNTS SET BALANCE = 150 WHERE ID = 1", ())?;
        tr_b.commit()?;
        println!("B committed BALANCE = 150");
    }

    let (still,): (i64,) = tr_a
        .query_first("SELECT BALANCE FROM ACCOUNTS WHERE ID = 1", ())?
        .unwrap();
    println!("snapshot still sees {}   <- stable view", still);
    tr_a.commit()?;

    let mut tr_rc = SimpleTransaction::new(&mut a, read_committed)?;
    let (now,): (i64,) = tr_rc
        .query_first("SELECT BALANCE FROM ACCOUNTS WHERE ID = 1", ())?
        .unwrap();
    println!("read committed sees {}   <- follows commits", now);
    tr_rc.commit()?;

    // -- 2. Conflict: two NO WAIT updates of the same row.
    let mut tr_a = SimpleTransaction::new(&mut a, snapshot)?;
    let mut tr_b = SimpleTransaction::new(&mut b, snapshot)?;
    tr_a.execute("UPDATE ACCOUNTS SET BALANCE = BALANCE + 1 WHERE ID = 1", ())?;
    println!("A updated the row (uncommitted)");
    match tr_b.execute("UPDATE ACCOUNTS SET BALANCE = BALANCE + 2 WHERE ID = 1", ()) {
        Ok(_) => println!("BUG: conflicting update succeeded"),
        Err(e) => println!("B's conflicting update failed as designed:\n    {}", e),
    }
    tr_b.rollback()?;
    tr_a.commit()?;

    println!("done.");
    Ok(())
}
